core/src/main/java/hudson/model/Executor.java

```diff
@@ -431,11 +431,12 @@ public SubTask call() throws Exception {
             } catch (AsynchronousExecution x) {
                 lock.writeLock().lock();
                 try {
-                    x.setExecutor(this);
+                    x.setExecutorWithoutCompleting(this);
                     this.asynchronousExecution = x;
                 } finally {
                     lock.writeLock().unlock();
                 }
+                x.maybeComplete();
```

Comment:
It's outside the `finally` block. May it cause issues, e.g. when an exception is thrown? One could argue that `lock.writeLock().unlock();`could also throw an exception in edge cases.

Reply:
Good point. Certainly the whole point is to move `x.maybeComplete()` to outside the locked region, but it could indeed be suppressed entirely if unlocking throws an exception (note that `setExecutorWithoutCompleting` will never throw).

I don't know what the correct behaviour is in such a case, and am happy to take advice. The javadoc for [`ReadWriteLock::writeLock`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReadWriteLock.html#writeLock()) and [`Lock::unlock`](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/Lock.html#unlock()) makes no mention of possible exceptions, but if you wanted to be safe I could do something like this (trying to propagate the correct exception, hence the complexity):

```java
            } catch (AsynchronousExecution x) {
                lock.writeLock().lock();
                RuntimeException orig = null;
                try {
                    x.setExecutorWithoutCompleting(this);
                    this.asynchronousExecution = x;
                } catch (RuntimeException e) {
                    orig = e;
                } finally {
                    try {
                        lock.writeLock().unlock();
                    } catch (RuntimeException e) {
                        if (orig == null) {
                            orig = e;
                        } else {
                            orig.addSuppressed(e);
                        }
                    } finally {
                        x.maybeComplete();
                    }
                }
                if (orig != null) {
                    throw orig;
                }
            }
```

Reply:
Yes, after some review I would prefer a more complex option. But I can live with the current impl as well. I was unable a case when it may happen on legitimate use-cases for the current lock class. OutOfMemory error does not count I'd guess

Seems @jglick is fine as well